src/metabase/native_query_analyzer.clj

```diff
@@ -37,50 +38,91 @@
-      (t2/select-pks-set :model/Field {:from [[(t2/table-name :model/Field) :f]]
-                                       :join [[(t2/table-name :model/Table) :t] [:= :table_id :t.id]]
-                                       :where [:and
-                                               [:= :t.db_id db-id]
-                                               (if (seq tables)
-                                                 [:in :%lower.t/name (map normalize-name tables)]
-                                                 true)
-                                               (if (seq columns)
-                                                 [:in :%lower.f/name (map normalize-name columns)]
-                                                 true)]}))
-    (catch JSQLParserException e
-      (log/error e "Error parsing native query"))))
+  [{:keys [columns tables] :as _parsed-query} db-id]
+  (t2/select-pks-set :model/Field (merge field-and-table-fragment
+                                         {:where [:and
+                                                  [:= :t.db_id db-id]
+                                                  (if (seq tables)
+                                                    [:in :%lower.t/name (map normalize-name tables)]
+                                                    ;; if we don't know what tables it's from, look everywhere
+                                                    true)
+                                                  (if (seq columns)
+                                                    [:in :%lower.f/name (map normalize-name columns)]
+                                                    ;; if there are no columns, it must be a select * or similar
+                                                    false)]})))
+
+(defn- indirect-field-ids-for-query
+  "Similar to direct-field-ids-for-query, but for wildcard selects"
+  [{:keys [table-wildcards has-wildcard? tables] :as _parsed-query} db-id]
+  (let [active-fields-from-tables
+        (fn [table-names]
+          (t2/select-pks-set :model/Field (merge field-and-table-fragment
+                                                 {:where [:and
+                                                          [:= :t.db_id db-id]
+                                                          [:= true :f.active]
+                                                          (if (seq table-names)
+                                                            [:in :%lower.t/name (map normalize-name table-names)]
+                                                            ;; if no tables, nothing is selectable
+                                                            false)]})))]
+
+    (if has-wildcard?
+      ;; select * from ...
+      ;; so, get everything in all the tables
+      (active-fields-from-tables tables)
+      ;; select foo.* from ..
+      ;; limit to the named tables
+      (active-fields-from-tables table-wildcards))))
```

Comment:
Wouldn't it be a little bit better if we had a `(cond has-wildcard? ... (seq tables) ... :else nil)`? So that 'nothing is selectable' is not even executed?

Reply:
yeah, that turned out nice. And `:else nil` is superfluous, and it let me save the `(seq table-names)` check in the helper fn.

Reply:
Yeah `:else nil` looks like `(if (= a true) false)`, sometimes your hands just want to type 🤣